dask_cuml/linear_model/linear_regression.py

```diff
@@ -0,0 +1,511 @@
+
+        """
+        Gather IPC handles for each worker and call _fit() on each worker
+        containing data.
+        """
+        exec_node = input_devarrays[0][0]
+
+        # Need to fetch parts on worker
+        on_worker = list(filter(lambda x: x[0] == exec_node, input_devarrays))
+        not_on_worker = list(filter(lambda x: x[0] != exec_node,
+                                    input_devarrays))
+
+        ipc_handles = [client.submit(get_input_ipc_handles, future,
+                                     workers=[a_worker])
+                       for a_worker, future in not_on_worker]
+
+        raw_arrays = [future for a_worker, future in on_worker]
+
+        # IPC Handles are loaded in separate threads on worker so they can be
+        # used to make calls through cython
+        # Calls _predict_on_worker defined in the bottom
+        ret = client.submit(_predict_on_worker, (ipc_handles, raw_arrays),
+                            self._build_params_map(), workers=[exec_node])
+
+        yield wait(ret)
+
+        dfs = [client.submit(series_on_worker, f, worker, loc_dict,
+                             X_df.npartitions, X_df, workers=[worker])
+               for worker, f in input_devarrays]
+
+        return dfs
+
+    def _build_host_dict(self, gpu_futures, client):
+
+        who_has = client.who_has(gpu_futures)
+
+        key_to_host_dict = {}
+        for key in who_has:
+            key_to_host_dict[key] = parse_host_port(who_has[key][0])
+
+        hosts_to_key_dict = {}
+        for key, host in key_to_host_dict.items():
+            if host not in hosts_to_key_dict:
+                hosts_to_key_dict[host] = set([key])
+            else:
+                hosts_to_key_dict[host].add(key)
+
+        workers = [key[0] for key in list(who_has.values())]
+        return build_host_dict(workers)
+
+
+def _fit_on_worker(data, params):
+    ipc_dev_list, devarrs_dev_list = data
+
+    # TODO: One ipc thread per device instead of per x,y,coef tuple
+    open_ipcs = []
+    for p, dev in ipc_dev_list:
+        for x, y, coef in p:
+            ipct = new_ipc_thread([x, y, coef], dev)
+            open_ipcs.append(ipct)
+
+    alloc_info = list(itertools.chain([t.info() for t in open_ipcs]))
+    alloc_info.extend(
+        list(itertools.chain(
+            [[(build_alloc_info(X)[0], build_alloc_info(y)[0],
+               build_alloc_info(coef)[0]) for X, y, coef in p]
+             for p, dev in devarrs_dev_list])))
+
+    print("ALLOC INFO FIT")
+    print(alloc_info)
+
+    try:
+        ols = cuOLS()
+
+        intercept = ols.fit_dask(alloc_info, params)
+
+    except Exception as e:
+        print("FAILURE in FIT: " + str(e))
+
+    intercept = 10.0
+
+    [t.close() for t in open_ipcs]
+    [t.join() for t in open_ipcs]
+
+    return intercept
+
+
+def _predict_on_worker(data, intercept, params=None):
+    ipc_dev_list, devarrs_dev_list = data
+
+    # TODO: One ipc thread per device instead of per x,y,coef tuple
```

Comment:
Since we're not able to use a threadpool here, we need to be able to bound the number of threads, otherwise this could get extremely out of control and hinder performance. We should be grouping by device here so that we have a hard-bound on the number of threads (which we know could never be a number larger than 16).